### PR TITLE
[SPARK-48511][SS] Remove TimeMode None from TransformWithState.

### PR DESCRIPTION
**Pull request by sahnib**
### What changes were proposed in this pull request?

This PR removes the TimeMode None from supported time modes in TransformWithState operator. A structured streaming query works in either Processing time mode, or event Time mode depending on whether eventTime has been specified, hence this change aligns TimeMode properly with Streaming query semantics. 

Note that if eventTimeColumn is specified for output dataset in transformWithState operator. TimeMode defaults to EventTime. 

### Why are the changes needed?

These changes are needed to align TimeMode values with how time flows in Streaming query. 

### Does this PR introduce _any_ user-facing change?

Yes, modifies the TimeMode semantic for transformWithState. 

### How was this patch tested?

All existing unit test pass. 

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Mind filing a JIRA please?

File: sql/api/src/main/java/org/apache/spark/sql/streaming/TimeMode.java [hunk 1]
```diff
@@ -31,11 +30,6 @@
 @Evolving
 public class TimeMode {
 
-    /**
-     * Neither timers nor ttl is supported in this mode.
-     */
-    public static final TimeMode None() { return NoTime$.MODULE$; }
```
Comment (HyukjinKwon) [inline sql/api/src/main/java/org/apache/spark/sql/streaming/TimeMode.java:37]: When did we add this? If this is already released out, we can't just remove and break it.

**Comment by sahnib** [comment]
> Mind filing a JIRA please?

Filed [SPARK-48511](https://issues.apache.org/jira/browse/SPARK-48511). Thanks.

File: sql/api/src/main/java/org/apache/spark/sql/streaming/TimeMode.java (hunk 1 above)
Comment (sahnib) [reply sql/api/src/main/java/org/apache/spark/sql/streaming/TimeMode.java:37]: We added this on April 7, 2024 as part of TTL support for Streaming State with TransformWithState operator. This operator is a new addition to Spark Structured Streaming, and is planned to be released in Spark 4.1. The API is currently in heavy development and evolving. I think its safe to make this change at this point, given its still unreleased in a Spark Major version.

**Review by anishshri-db — APPROVED** [review]

File: sql/api/src/main/java/org/apache/spark/sql/streaming/TimeMode.java (hunk 1 above)
Comment (anishshri-db) [reply sql/api/src/main/java/org/apache/spark/sql/streaming/TimeMode.java:37]: yea not released yet. This is only in the 4.0 branch. 
@sahnib - while we are here - is there a way to make this package private to `sql` too on the Java side ?

**Comment by sahnib** [comment]
@HeartSaVioR PTAL, thanks.

File: sql/api/src/main/java/org/apache/spark/sql/streaming/TimeMode.java (hunk 1 above)
Comment (HeartSaVioR) [reply sql/api/src/main/java/org/apache/spark/sql/streaming/TimeMode.java:37]: Java doesn't support package private with specifying ancestor package. It's always limited to the same package and it won't work for our case.
(It might change in recent versions though I haven't heard about such change.)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala [hunk 2]
```diff
@@ -502,6 +502,6 @@ class IncrementalExecution(
       case p: StateStoreWriter => p.shouldRunAnotherBatch(
         watermarkPropagator.getInputWatermarkForEviction(tentativeBatchId,
           p.stateInfo.get.operatorId))
-    }.exists(_ == true)
+    }.contains(true)
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala:505]: nit: unrelated change? Please file a new minor PR.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithListStateSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithListStateSuite.scala:1]: I wonder why the test code change is required given the code change is just to remove TimeMode None. Do we fix some test flakiness as well here, or what is the rationale of manual clock?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala:1]: Same question.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala:1]: Same.

**Review by HeartSaVioR — COMMENTED** [review]
Thanks for the contribution. Left comments. General comments are 1) whether we can use any time mode where we could use None, or only ProcessingTime, and 2) why we need manual clock while we just removed out time mode None (doesn't sound to be relevant).

**Comment by HeartSaVioR** [comment]
Also let's revisit the UX. If they use neither timeout nor TTL, they could just do None regardless they have event time column or not. Given we remove None, what is the expectation of UX? Does user have to specify either one based on what they have, even though they never use timeout? If this is the case, that doesn't sound like a better UX.

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java [hunk 3]
```diff
@@ -206,7 +206,7 @@ public void testInitialStateForTransformWithState() {
 
     Dataset<String> transformWithStateMapped = grouped.transformWithState(
       new TestStatefulProcessorWithInitialState(),
-      TimeMode.None(),
+      TimeMode.ProcessingTime(),
```
[the remark this answers is not in the document]
Comment (sahnib) [reply sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java:209]: either ProcessingTime or EventTime works as a replacement.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala (hunk 2 above)
Comment (sahnib) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala:505]: will remove these minor changes from this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ExpiredTimerInfoImpl.scala [hunk 4]
```diff
@@ -27,8 +27,7 @@ import org.apache.spark.sql.streaming.{ExpiredTimerInfo, TimeMode}
  */
 class ExpiredTimerInfoImpl(
     isValid: Boolean,
-    expiryTimeInMsOpt: Option[Long] = None,
-    timeMode: TimeMode = TimeMode.None()) extends ExpiredTimerInfo {
```
[the remark this answers is not in the document]
Comment (sahnib) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ExpiredTimerInfoImpl.scala:31]: That's a good question. `TimeMode` is passed in the StatefulProcessor.init function to the user. We were also passing it to the  `ExpiredTimerInfoImpl` but the interface `ExpiredTimerInfo` does not even expose it to the user, and we never set this variable with proper timeMode. Hence, I have removed it. I dont think this was intended to here. cc: @anishshri-db can you confirm?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala [hunk 5]
```diff
@@ -632,9 +632,9 @@ class MicroBatchExecution(
     // Check whether next batch should be constructed
     val lastExecutionRequiresAnotherBatch = noDataBatchesEnabled &&
       // need to check the execution plan of the previous batch
-      execCtx.previousContext.map { plan =>
+      execCtx.previousContext.exists { plan =>
```
[the remark this answers is not in the document]
Comment (sahnib) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:635]: will do.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ListStateSuite.scala [hunk 6]
```diff
@@ -38,7 +38,7 @@ class ListStateSuite extends StateVariableSuiteBase {
     tryWithProviderResource(newStoreProviderWithStateVariable(true)) { provider =>
       val store = provider.getStore(0)
       val handle = new StatefulProcessorHandleImpl(store, UUID.randomUUID(),
-        Encoders.STRING.asInstanceOf[ExpressionEncoder[Any]], TimeMode.None())
+        Encoders.STRING.asInstanceOf[ExpressionEncoder[Any]], TimeMode.ProcessingTime())
```
[the remark this answers is not in the document]
Comment (sahnib) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ListStateSuite.scala:41]: just a convenience.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithListStateSuite.scala
Comment (sahnib) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithListStateSuite.scala:1]: `TransformWithStateExec` currently returns `true` for ProcessingTime timeMode in `shouldRunAnotherBatch`. (Note that this behavior is same as FMGWS). I had to add a manual clock to ensure that data processing is triggerred before Check/Expect blocks (otherwise AddData does not finish). Let me know if that answers your question.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala
Comment (sahnib) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala:1]: https://github.com/apache/spark/pull/46825#discussion_r1636698711

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala
Comment (sahnib) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala:1]: https://github.com/apache/spark/pull/46825#discussion_r1636698711

**Comment by sahnib** [comment]
> Also let's revisit the UX. If they use neither timeout nor TTL, they could just do None regardless they have event time column or not. Given we remove None, what is the expectation of UX? Does user have to specify either one based on what they have (event time column is set or not), even though they never use timeout? If that is the case, that doesn't sound like a better UX.

Even with TimeMode None -> the query has eventTime (if eventTimeColumn is specified), and processingTime (based on driver's clock). I think this makes TimeMode None confusing. 
As a general rule of thumb, customer should set TimeMode as EventTime if eventTimeColumn is specified, else use ProcessingTime.

**Comment by HeartSaVioR** [comment]
Just leaving history here: currently processing time mode triggers the batch continuously which is not acceptable if there is no timer/TTL to check. We are now discussing how to handle the case.
